File: backend/schemas/purchase_order.py
# backend/schemas/purchase_order.py
from pydantic import BaseModel, Field, ConfigDict
from typing import List, Optional, Dict, Any
from datetime import datetime
import uuid
from models.purchase_order import PurchaseOrderStatus
from models.dealer import Dealer as DealerModel

class DocumentSchema(BaseModel):
    document_id: uuid.UUID
    document_name: str
    document_type: str
    document_url: str
    

class ProductBase(BaseModel):
    product_id: uuid.UUID
    name: str
    pack_size: Optional[str] = None
    trade_price_incl_vat: float
    mrp: Optional[float] = None

    class Config:
        orm_mode = True
        populate_by_name = True

class PurchaseOrderItemBase(BaseModel):
    product_id: uuid.UUID
    quantity: int = Field(..., gt=0)

class PurchaseOrderItemCreate(PurchaseOrderItemBase):
    pass

class PurchaseOrderItem(BaseModel):
    po_item_id: int
    po_id: int
    product_id: uuid.UUID
    quantity: int
    pack_size_snapshot: Optional[str] = None
    unit_price: float
    total_price: float
    product: Optional[ProductBase] = None

    class Config:
        orm_mode = True

class PurchaseOrderBase(BaseModel):
    pass

class PurchaseOrderCreate(BaseModel):
    dealer_id: uuid.UUID
    items: List[PurchaseOrderItemCreate]

class PurchaseOrderUpdate(BaseModel):
    items: List[PurchaseOrderItemCreate]

class DealerBase(BaseModel):
    dealer_id: uuid.UUID
    company_name: str
    contact_person: str
    contact_number: str
    email: Optional[str] = None
    billing_address: str
    shipping_address: str

    class Config:
        orm_mode = True

class PurchaseOrder(BaseModel):
    po_id: int
    po_number: str
    dealer_id: uuid.UUID
    dealer: Optional[DealerBase] = None
    created_by_user: uuid.UUID
    po_date: datetime
    status: PurchaseOrderStatus
    total_tp: Optional[float] = None
    total_vat: Optional[float] = None
    approved_at: Optional[datetime] = None
    # Calculated fields (not in DB, computed from items)
    total_ex_vat: Optional[float] = None
    vat_percent: Optional[float] = None
    vat_amount: Optional[float] = None
    total_inc_vat: Optional[float] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    items: List[PurchaseOrderItem] = []

    class Config:
        orm_mode = True


class PurchaseOrderList(BaseModel):
    """Schema for a list of purchase orders with pagination info."""
    items: List[PurchaseOrder]
    total: int
    skip: int
    limit: int

    class Config:
        orm_mode = True

File: frontend/src/types/purchaseOrder.ts
// frontend/src/types/purchaseOrder.ts

export interface ProductDetails {
  product_id: string;
  name: string;
  pack_size: string;
  trade_price_incl_vat: number;
  mrp: number;
  stock_qty: number;
  sku_code: string;
  status: string;
}

export interface PurchaseOrderItem {
  po_item_id: number;
  po_id: number;
  product_id: string;
  product: ProductDetails;  // Made required since we always expect product details
  pack_size_snapshot: string;
  quantity: number;
  unit_price: number;
  total_price: number;
}

export interface DealerDetails {
  dealer_id: string;
  company_name: string;
  contact_person: string;
  contact_number: string;
  email?: string;
  billing_address: string;
  shipping_address: string;
  created_at: string;
  updated_at?: string;
}

export interface PurchaseOrder {
  po_id: number;
  po_number: string;
  dealer_id: string;
  dealer: DealerDetails;  // Always included in the response
  created_by_user: string;
  external_ref_code: string | null;
  po_date: string;
  status: 'draft' | 'submitted' | 'approved' | 'invoiced' | 'cancelled';
  total_ex_vat: number;
  vat_percent: number;
  vat_amount: number;
  total_inc_vat: number;
  approved_at: string | null;
  created_at: string;
  updated_at: string | null;
  items: PurchaseOrderItem[];  // Always included with product details
  combined_po_id: number | null;
}

export interface PurchaseOrderItemCreate {
  product_id: string;
  quantity: number;
  unit_price: number;
}

export interface PurchaseOrderCreate {
  dealer_id: string;
  external_ref_code?: string;
  items: PurchaseOrderItemCreate[];
}
